src/main/scala/com/gu/fastly/Lambda.scala

```diff
@@ -56,19 +66,51 @@ class Lambda {
     dotcomSurrogateKey
   }
 
-  private def sendFastlyPurgeRequestAndAmpPingRequest(contentId: String, purgeType: PurgeType, serviceId: String, surrogateKey: String, fastlyApiKey: String): Boolean = {
-    if (sendFastlyPurgeRequest(contentId, purgeType, serviceId, surrogateKey, fastlyApiKey))
-      sendAmpPingRequest(contentId)
-    else
-      false
+  // Left[Id] for Delete, Right[Id] for Update
+  def decidePurgeType: PartialFunction[Event, Either[Id, Id]] = {
+    case Event(_, EventType.Update, _, _, Some(EventPayload.Content(content))) => Right(content.id)
+    case Event(_, EventType.RetrievableUpdate, _, _, Some(EventPayload.RetrievableContent(content))) => Right(content.id)
+    case Event(_, EventType.Delete, _, _, Some(EventPayload.RetrievableContent(content))) => Left(content.id)
+  }
+
+  def purgeDelete: Id => Future[List[PurgeResult]] = { id =>
+    val fastlyResult: Future[Boolean] = sendFastlyPurgeRequest(id, Hard, config.fastlyDotcomServiceId, makeDotcomSurrogateKey(id), config.fastlyDotcomApiKey)
+    val ampResult: Future[Boolean] = fastlyResult.flatMap(if (_) sendAmpPingRequest(id) else Future.successful(false))
+    Future.sequence(fastlyResult.map(PurgeResult(_, Dotcom)) :: ampResult.map(PurgeResult(_, AMP)) :: Nil)
+  }
+
+  def purgeUpdate: Id => Future[List[PurgeResult]] = { id =>
+    Future.sequence(
+      sendFastlyPurgeRequest(id, Soft, config.fastlyDotcomServiceId, makeDotcomSurrogateKey(id), config.fastlyDotcomApiKey).map(PurgeResult(_, Dotcom)) ::
+        sendFastlyPurgeRequest(id, Soft, config.fastlyApiNextgenServiceId, makeDotcomSurrogateKey(id), config.fastlyDotcomApiKey).map(PurgeResult(_, Nextgen)) ::
+        sendFastlyPurgeRequest(id, Soft, config.fastlyMapiServiceId, makeMapiSurrogateKey(id), config.fastlyMapiApiKey).map(PurgeResult(_, Mapi)) ::
+        Nil
+    )
+  }
+
+  def log: List[(Id, List[PurgeResult])] => Unit = { results =>
+    val purgeResultCounts = results.flatMap(_._2).groupBy(identity).mapValues(_.size)
+    purgeResultCounts.foreach {
+      case (PurgeResult(true, service), idCount) => println(s"Successfully purged $service for $idCount content ids")
+      case (PurgeResult(false, service), idCount) => println(s"Failed to purge $service for $idCount content ids")
+    }
+  }
+
+  def idForFullySuccessfulPurge: PartialFunction[(Id, List[PurgeResult]), Id] = {
```

Comment:
would it be clearer to have this inline rather than as a separate PartialFunction since it's only used once?

Reply:
Nice idea